glean-core/src/event_database/mod.rs

```diff
@@ -0,0 +1,325 @@
+        extra: Option<HashMap<String, String>>,
+    ) {
+        let event = RecordedEventData {
+            timestamp,
+            category: meta.category.to_string(),
+            name: meta.name.to_string(),
+            extra,
+        };
+        let event_json = serde_json::to_string(&event).unwrap();
+        let mut stores_to_send: Vec<&str> = Vec::new();
+
+        {
+            let mut db = self.event_stores.write().unwrap();
+            for store_name in meta.send_in_pings.iter() {
+                let store = db
+                    .entry(store_name.to_string())
+                    .or_insert_with(|| Vec::new());
+                store.push(event.clone());
+                self.write_event_to_disk(store_name, &event_json);
+                if store.len() == glean.get_max_events() {
+                    stores_to_send.push(&store_name);
+                }
+            }
+        }
+
+        for store_name in stores_to_send {
+            if let Err(err) = glean.send_ping_by_name(store_name, false) {
+                log::error!(
+                    "Got more than {} events, but could not send {} ping: {}",
+                    glean.get_max_events(),
+                    store_name,
+                    err
+                );
+            }
+        }
+    }
+
+    /// Writes an event to a single store on disk.
+    ///
+    /// # Arguments
+    ///
+    /// * `store_name` - The name of the store.
+    /// * `event_json` - The event content, as a single-line JSON-encoded string.
+    fn write_event_to_disk(&self, store_name: &str, event_json: &str) {
+        let _lock = self.file_lock.write().unwrap();
+        if let Err(err) = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(self.path.join(store_name))
+            .and_then(|mut file| writeln!(file, "{}", event_json))
+        {
+            log::error!("Error writing event to store {}: {}", store_name, err);
+        }
+    }
+
+    /// Get a snapshot of the stored event data as a JsonValue.
+    ///
+    /// # Arguments
+    ///
+    /// * `store_name` - The name of the desired store.
+    /// * `clear_store` - Whether to clear the store afterward.
+    ///
+    /// # Returns
+    ///
+    /// The an array of events, JSON encoded, if any.
+    pub fn snapshot_as_json(&self, store_name: &str, clear_store: bool) -> Option<JsonValue> {
+        let result = {
+            let db = self.event_stores.read().unwrap();
+            db.get(&store_name.to_string()).and_then(|store| {
+                if !store.is_empty() {
+                    let first_timestamp = store[0].timestamp;
+                    Some(json!(store
+                        .iter()
+                        .map(|e| e.serialize_relative(first_timestamp))
+                        .collect::<Vec<JsonValue>>()))
+                } else {
+                    None
+                }
+            })
+        };
+
+        if clear_store {
+            {
```

Comment:
These double-nested blocks look weird.
It's just to shorten the scope of the locks, right?

Reply:
(which means we can at least leave out the 2nd block?

Reply:
Exactly.  The first block actually can be converted to a single statement.  And then, yes, the second is redundant because it's the last block in the outer block.  So, can be much simplified.